src/blocks/header.rs

```diff
@@ -0,0 +1,333 @@
+            return Err(Error::Validation(format!(
+                "expected final beacon entry in block to be at round {}, got: {}",
+                max_round,
+                last.round()
+            )));
+        }
+
+        let mut prev = prev_entry;
+        for curr in &self.beacon_entries {
+            if !curr_beacon
+                .verify_entry(curr, prev)
+                .map_err(|e| Error::Validation(e.to_string()))?
+            {
+                return Err(Error::Validation(format!(
+                    "beacon entry was invalid: curr:{curr:?}, prev: {prev:?}"
+                )));
+            }
+            prev = curr;
+        }
+        Ok(())
+    }
+
+    /// Serializes the header to bytes for signing purposes i.e. without the
+    /// signature field
+    fn signing_bytes(&self) -> Vec<u8> {
+        let mut blk = self.clone();
+        blk.signature = None;
+        fvm_ipld_encoding::to_vec(&blk).expect("block serialization cannot fail")
+    }
+}
+
+/// A [`RawBlockHeader`] which caches calls to [`RawBlockHeader::cid`] and [`RawBlockHeader::verify_signature_against`]
+#[derive(Debug, Default)]
+pub struct CachingBlockHeader {
+    uncached: RawBlockHeader,
+    cid: OnceCell<Cid>,
+    has_ever_been_verified_against_any_signature: AtomicBool,
+}
+
+impl PartialEq for CachingBlockHeader {
+    fn eq(&self, other: &Self) -> bool {
+        self.cid() == other.cid()
+    }
+}
+
+impl Clone for CachingBlockHeader {
+    fn clone(&self) -> Self {
+        Self {
+            uncached: self.uncached.clone(),
+            cid: self.cid.clone(),
+            has_ever_been_verified_against_any_signature: AtomicBool::new(
+                self.has_ever_been_verified_against_any_signature
+                    .load(Ordering::Acquire),
+            ),
+        }
+    }
+}
+
+impl Deref for CachingBlockHeader {
+    type Target = RawBlockHeader;
+
+    fn deref(&self) -> &Self::Target {
+        &self.uncached
+    }
+}
+
+impl CachingBlockHeader {
+    pub fn new(uncached: RawBlockHeader) -> Self {
+        Self {
+            uncached,
+            cid: OnceCell::new(),
+            has_ever_been_verified_against_any_signature: AtomicBool::new(false),
+        }
+    }
+    pub fn into_raw(self) -> RawBlockHeader {
+        self.uncached
+    }
+    /// Returns [`None`] if the blockstore doesn't contain the CID.
+    pub fn load(store: impl Blockstore, cid: Cid) -> anyhow::Result<Option<Self>> {
+        if let Some(uncached) = store.get_cbor::<RawBlockHeader>(&cid)? {
+            Ok(Some(Self {
+                uncached,
+                cid: OnceCell::with_value(cid),
+                has_ever_been_verified_against_any_signature: AtomicBool::new(false),
+            }))
+        } else {
+            Ok(None)
+        }
+    }
+    pub fn cid(&self) -> &Cid {
+        self.cid.get_or_init(|| self.uncached.cid())
+    }
+
+    pub fn verify_signature_against(&self, addr: &Address) -> Result<(), Error> {
```

Comment:
I don't understand why this is OK, but if it isn't that belongs in a separate PR